### DISCUSSION TITLE
A moldable object inspector

### DISCUSSION THREAD

**Discussion by khinsen** [rabbibotton/clog#327] [Show and tell]
Here is my first CLOG-based project: a moldable object inspector

https://codeberg.org/khinsen/clog-inspector

It's heavily inspired by the Smalltalk object inspector in [Glamourous Toolkit](https://gtoolkit.com/). The focus is on making it very cheap to add new views to the inspector, in order to encourage developers to make views for their objects from the earliest stages of development.

Thanks @rabbibotton for CLOG, I couldn't have done this without it and exploring CLOG is a pleasure due to the readable code and complete documentation!

**Comment by aykaramba** [comment]
Cool!  That looks really interesting, I will be playing with this tonight.  Very exciting times in the Lisp world, it seems.

**Comment by rabbibotton** [comment]
Just played with it a bit, really looking forward to your work!

**Comment by rabbibotton** [comment]
I would consider suggesting people use CLOG and CLOG-ACE from ultralisp as that will always have the latest changes for both.

**Reply by rabbibotton** [reply]
your readme suggested quicklisp etc

**Reply by khinsen** [reply]
Thanks for the suggestion! I have to admit that I have never used Ultralisp myself. It's been on my to-do list for... a long time. I just put clog and clog-ace into the local-projects directory of Quicklisp.

**Reply by khinsen** [reply]
So I had a quick look at Ultralisp. It turned out to be easier to submit a package than to understand how Ultralisp works (given that Quicklisp has no documentation on additional distributions), so I started by submitting clog-moldable-inspector as a first experiment :-)
If all goes well, it will be in the next build of Ultralisp.

And to improve my understanding of Quicklisp, I started writing a couple of inspector views for it.

**Comment by khinsen** [comment]
Latest news: there's a 9 minute demo video of my inspector: https://diode.zone/w/cd5s1yd1iJrhiMR1jS1xz3

**Comment by rabbibotton** [comment]
looking now :)

**Comment by rabbibotton** [comment]
Really love the idea of your project. I think will try organizing
properties in the GUI builder with same principles. Are you open to
switching to a BSD/MIT sort license if I  see ways to integrate your work
directly? I have no interest in making my work gpl which would be required
to use it.

**Reply by khinsen** [reply]
Any OSI licence is fine with me!

**Reply by tluyben** [reply]
This is a very exciting feature; I believe in easily creating views / configuration tools for dev all the time. Of course in vs code you can do these things, but in smalltalk (but also delphi for instance, to a certain extend) it's very fast and easy to do and it makes sense it also is in CL. Allowing to have full control over your code, the editor, but also the entire IDE is a very powerful thing and allows for easier configuration between juniors/seniors/newcomers etc if it's well integrated into the entire concept. 

Hope the existing work can be integrated and extended.

**Reply by khinsen** [reply]
@rabbibotton I just changed the license to 3-clause BSD.

**Reply by rabbibotton** [reply]
Awesome

**Reply by rabbibotton** [reply]
I have a few things to finish up still in my 2.1 target on the general IDE side, once done will be working on the GUI builder side other areas that I want to look at new approaches beyond what already exists for CL and I see this or something modeled on it being a part of it for sure.

**Comment by rabbibotton** [comment]
@khinsen you should look at the new clog-break you can insert your inspector in the :run

**Reply by khinsen** [reply]
Interesting, thanks for the pointer!

It looks like I cannot pass any arguments to `run`, so how could I pick the object I want to inspect?

**Comment by rabbibotton** [comment]
clog-break takes a form for :run that on break runs (I realize now would be
good to use the return value to decide if to still promote to continue
execution) so you would do you inspect there in the lexicle scope of the
clog-break of that form

**Reply by rabbibotton** [reply]
A clog-obj is passed (a clog-body object usually) so that one has access to the browser where the debug info is being displayed (not per se same machine even) for example you could use:

(dotimes (n 100) (when (eql n 1) (clog-break :run (lambda (obj) (inspect n))) (print n) (sleep .5))

**Reply by rabbibotton** [reply]
added: In clog-break if the form passed to run returns :continue the break confirm dialog is not displayed and execution continues

**Reply by khinsen** [reply]
That's all very nice, thanks!

**Reply by rabbibotton** [reply]
BTW one thing that is worth while knowing, say you have a text or gtk app etc not clog based, you can use clog-break also (you don't have to be using the builder for clog-break or clog-probe) you just run the clog-repl with:

(clog-user:clog-repl :use-clog-debugger t)

Then clog-break and and clog-probe will use that repl window for debug.

**Reply by khinsen** [reply]
That was the missing piece for me, thanks!

**Reply by rabbibotton** [reply]
Looking forward to see some working demos reals on reddit from you :P show your stuff off it is awesome!

**Comment by tluyben** [comment]
Related ideas: 

https://x.com/feenkcom/status/1797700783628259772

**Comment by rabbibotton** [comment]
Let me know what other hooks into builder would make sense to add menus etc. I imagine we should add hooks into the project panel and dir panel so that file names can be sent to the  clog-moldable-inspector

**Reply by khinsen** [reply]
Files and directories is a big one. I am starting to add views for `clog-element` and its subclasses. I'd expect them to be useful for the builder as well, but they are still rather primitive at this time.

**Comment by rabbibotton** [comment]
I added the ability to add extensions to handle files and directories, I opened a ticket on your project with code to add

**Comment by rabbibotton** [comment]
Here is pic of using the CLOG Probe Panel ability to add probes from an eval  and CMI launched from the panel on a CLOG-OBJ, you may find this useful.

<img width="1241" alt="image" src="https://github.com/rabbibotton/clog/assets/11616789/1ca60f42-ffb4-47f8-930f-f72912ba0455">

**Reply by khinsen** [reply]
Nice!

I see you have my latest additions from yesterday. Have a look at the "Operations" view. It has all methods that specialize on clog-element and can be called with a single argument. By clicking on the arrow to the left, you can run the method and inspect the result. Very useful for looking at the many attributes of a clog-element.

What I plan to improve is the presentation of the operations. Group them by package, for example. And add a per-class filter. For clog-element, I'd remove the create- methods because they are not useful in the inspector.

The HTML view needs some pretty-printing to be useful. That's why I added the HTML/sexp view as a workaround, profiting from CL's built-in pretty-printing for s-expressions.

**Comment by khinsen** [comment]
There's now a built-in tutorial:
```
(clog-moldable-inspector:tutorial)
```
